Cloudstack Driver - Create/Delete Network functionality + tests

boul opened the pull request:
Functions added on the cloudstack driver:

ex_list_network_offering
ex_create_network
ex_delete_network

New Object:
CloudStackNetworkOffering

And some additions to the CloudStackNetwork dictionary
netmask
gatway
vpcid

Commit:
Add support for network creation on Cloudstack advanced zones

Commit:
test for ex_list_networkofferings

Commit:
fix long lines

Commit:
add delete_network function and tests

Kami commented:
/cc @runseb

Kami commented on libcloud/compute/drivers/cloudstack.py:
```diff
@@ -453,6 +465,30 @@ def __repr__(self):
+    Class representing a CloudStack Network Offering.
+    """
+
+    def __init__(self, name, displaytext, guestiptype, id, serviceofferingid,
```
For consistency with existing code and other drivers, please use `_` to separate words in the argument names - display_text, guest_ip_type, service_offerring_id, for_vpc, etc.

Kami commented on libcloud/compute/drivers/cloudstack.py:
```diff
@@ -880,6 +916,122 @@ def ex_list_networks(self):
+
+        return networkofferings
+
+    def ex_create_network(self, displaytext, name, networkoffering,
```
Same here, please use underscore in the argument names.

Kami replied:
Instead of using kwargs, please explicitly declare `gateway`, `netmask` and `vpc_id` argument in the method signature.

Using kwargs in cases like this makes programmatic introspection and many other things hard / impossible - https://libcloud.readthedocs.org/en/latest/development.html#don-t-abuse-kwargs

sebgoa replied:
Could you also add deleteNetworkOffering and createNetworkOffering ? Since you are defining the CloudStackNetworkOffering class, you could add the other methods

boul replied:
createNetworkOffering and deleteNetworkOffering are only-available to ROOT-admins. I might add this at a later stage. Especially createNetworkOffering could be tricky based on the different network providers. I'd rather put some focus on CRUD VPC's.

Kami replied:
Yeah, and it's also worth noting that compute drivers should only implement user / non-admin functionality.

We need to find a better and unified solution (maybe a separate driver / class) for handling admin actions, but this requires a bit more thought and work.

sebgoa replied:
On Jun 11, 2014, at 3:56 PM, Tomaz Muraus notifications@github.com wrote:

> In libcloud/compute/drivers/cloudstack.py:
> 
> > -        networkofferings = []
> >   +
> > -        for netoffer in netoffers:
> >   +
> > -            networkofferings.append(CloudStackNetworkOffering(
> > -                            netoffer['name'],
> > -                            netoffer['displaytext'],
> > -                            netoffer['guestiptype'],
> > -                            netoffer['id'],
> > -                            netoffer['serviceofferingid'],
> > -                            netoffer['forvpc'],
> > -                            self))
> >   +
> > -        return networkofferings
> >   +
> > -    def ex_create_network(self, displaytext, name, networkoffering,
> 
> Yeah, and it's also worth noting that compute drivers should only implement user / non-admin functionality.
> 
> We need to find a better and unified solution (maybe a separate driver / class) for handling admin actions, but this requires a bit more thought and work.

Yes agreed, good point by Roland.

> —
> Reply to this email directly or view it on GitHub.

Kami commented on libcloud/compute/drivers/cloudstack.py:
```diff
@@ -880,6 +916,122 @@ def ex_list_networks(self):
+
+        return network
+
+    def ex_delete_network(self, network, forced=None):
```
For consistency with existing code base, please call the argument `force` instead of `forced`.

Commit:
fix indents

Commit:
minor changes based on pull request feedback

boul commented:
Thx for the quick feedback, see 50e6c75 with the suggested changes for consistancy reasons.

Commit:
fix long line

sebgoa commented on libcloud/compute/drivers/cloudstack.py:
```diff
@@ -880,6 +917,122 @@ def ex_list_networks(self):
+        return networkofferings
+
+    def ex_create_network(self, displaytext, name, networkoffering,
+                          location, **kwargs):
```
You still need to fix the kwargs issue. We prefer to explicitly list the arguments that you need: gateway, net mask, vpc_id

boul replied:
Fixed it. Cheers!

Commit:
getting rid of kwargs,
add optinal args for ex_create_network
extend network dictionary
fix tests accordingly

Commit:
more tests

Commit:
fix whitespace and comments

sebgoa commented:
Can you squash your commits ? that would make a clean PR. thx

Kami commented:
I've squashed the commits, renamed some variables to use underscores and merged patch into trunk. Thanks.

boul commented:
Thx just squashed before noticing you already did. Thx!

Kind Regards,
Roeland Kuipers

http://www.schubergphilis.com

Sent from a mobile device.

> Op 12 jun. 2014 om 20:51 heeft Tomaz Muraus notifications@github.com het volgende geschreven:
> 
> I've squashed the commits, renamed some variables to use underscores and merged patch into trunk. Thanks.
> 
> —
> Reply to this email directly or view it on GitHub.

sebgoa commented:
@Kami you were not that nice with me when I submitted pr :) …maybe once

Kami commented:
@runseb heh, I can't always easily squash the commits - if the branch is not in sync with trunk, trying to apply the patch results in conflicts and resolving this is quite painful (although, I do it every now and then).
